[5.3] Queued session garbage collection

HolgerW1 opened the pull request:
See #13096

Session garbage collections are currently executed as part of the standard request cycle. This may have negative impacts on requests.

If a garbage collection takes a long time then the request of a user is unnecessarily delayed for no obvious reason (at least from a user's perspective). This might even lead to a request timeout.
Furthermore if a gc throws an exception the whole request will fail even though the gc doesn't have anything to do with the request the user made. If the same request is made again by the user it will most likely not fail which would be very confusing.

Queueing gc operations and processing them outside of a request cycle makes more sense and does make the whole system more robust.

Commit:
Queue garbage collection operations instead of executing them as part of the request

Commit:
Removed all unnecessary dependencies
Use dependency injection for the job dispatcher

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,44 @@
+     * Create a new job instance.
+     *
+     * @param \SessionHandlerInterface $sessionHandler
+     */
```
missing param annotation and missing return void

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * Create a new job instance.
+     *
+     * @param \SessionHandlerInterface $sessionHandler
```
incorrect

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,44 @@
+namespace Illuminate\Session;
+
+use Illuminate\Contracts\Queue\ShouldQueue;
+use SessionHandlerInterface;
```
?

GrahamCampbell replied:
this isn't the correct class name

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -32,11 +41,13 @@ class StartSession
      * @return void
      */
-    public function __construct(SessionManager $manager)
+    public function __construct(SessionManager $manager, DispatcherContract $dispatcher)
```
please allow this to be null

GrahamCampbell replied:
as in: `, DispatcherContract $dispatcher = null)`, and update the phpdoc to add `|null`.

GrahamCampbell commented on src/Illuminate/Session/Middleware/StartSession.php:
```diff
@@ -149,7 +160,7 @@ protected function collectGarbage(SessionInterface $session)
         // hit it, we'll call this handler to let it delete all the expired sessions.
         if ($this->configHitsLottery($config)) {
-            $session->getHandler()->gc($this->getSessionLifetimeInSeconds());
+            $this->dispatcher->dispatch(new CollectGarbageJob($session->getHandler(), $this->getSessionLifetimeInSeconds()));
```
please update this to allow the dispatcher to be null

GrahamCampbell replied:
```
if ($this->dispatcher) {
    $this->dispatcher->dispatch(new CollectGarbageJob($session->getHandler(), $this->getSessionLifetimeInSeconds()));
} else {
    $session->getHandler()->gc($this->getSessionLifetimeInSeconds());
}
```

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * The session handler instance for which the garbage collection should run.
+     *
+     * @var \SessionHandlerInterface
```
no such interface exists?

HolgerW1 replied:
What do you mean?
http://php.net/manual/en/class.sessionhandlerinterface.php

Commit:
phpdoc fixed
Allow dispatcher to be null and use a direct call for the garbage collection instead of a queue

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,46 @@
+namespace Illuminate\Session;
+
+use Illuminate\Contracts\Queue\ShouldQueue;
+use SessionHandlerInterface;
```
Please sort imports by length

GrahamCampbell commented on src/Illuminate/Session/CollectGarbageJob.php:
```diff
@@ -0,0 +1,46 @@
+     * @param  int  $sessionLifetime
+     * @return void
+     */
+    public function __construct(SessionHandlerInterface $sessionHandler, $sessionLifetime)
```
naming these simply `$handler` and `$lifetime` would be preferable

GrahamCampbell replied:
same goes for the matching properties

Commit:
Include order
Briefer naming of properties

taylorotwell commented:
I don't really like this. I would suggest just using a cache based handler that doesn't need garbage collection such as Redis or Memcached.

HolgerW1 commented:
> I don't really like this.

Could you provide further details why? The current implementation is fragile and error-prone as described above and I can't see any reason why the GC of foreign sessions should take place inside the request cycle.
